tests/lax_numpy_test.py

```diff
@@ -1072,6 +1072,23 @@ def test_single(m, args_maker, repeats, axis):
     for repeats in [2, [1,3,2], [2], lnp.array([1,3,2]), lnp.array([2])]:
       test_single(m_rect, args_maker, repeats, axis=1)
 
+  def testIssue2330(self):
+    '''
+    Make sure return value of jnp.concatenate is a jax.ndarray and is side-effect save
+    '''
+    def attempt_sideeffect(x):
+      x = [x]
+      x = jnp.concatenate(x)
+      x -= 1.
+    return x
+
+    np_input = np.ones((1))
+    jnp_input = jnp.ones((1))
+    attempt_sideeffect(np_input)
+    attempt_sideeffect(jnp_input)
+
+    self.assertAllClose(np_input, jnp_input, check_dtypes=False)
```

Comment:
Since you modify both the `expected` and `actual` arrays here it's not 100% clear whether it's doing a side-effect on both or not doing a side effect.

I think it would be clearer to have an unmodified array for `expected` to compare against.

Also, it would be nice to have a very minimal check just verifying that the result of `jnp.concatenate` on a single array is JAX `DeviceArray`. Honestly just checking that `type(jnp.concatenate([numpy_array])) is DeviceArray` might be sufficient, without bothering to do the full integration test here.

Reply:
> Since you modify both the expected and actual arrays here it's not 100% clear whether it's doing a side-effect on both or not doing a side effect.

True, though you could say that this test is checking the behavior is consistent for both types of input, regardless of what the result actually is. Either way seems fine to me.

> Also, it would be nice to have a very minimal check just verifying that the result of jnp.concatenate on a single array is JAX DeviceArray. 

I like the integration test because once a bug is observed, especially one that came up in the wild like this one, we want to never regress it, and the best way to do that is to test it directly.

Luckily, with tests we don't need to think "either/or", since we can do both! It could be a separate test method, or it could just be a `self.assertIsInstance(x, DeviceArray)` in the body of `attempt_sideeffect` here. @msbauer up to you how/whether to add that kind of additional check.